src/NuGet.Core/NuGet.Resolver/ResolverUtility.cs

```diff
@@ -334,50 +334,125 @@ internal static bool ShouldRejectPackagePair(ResolverPackage p1, ResolverPackage
         /// </summary>
         public static IEnumerable<ResolverPackage> FindCircularDependency(IEnumerable<ResolverPackage> solution)
         {
+            // to keep track of visited packages to avoid processing them again
+            var visitedPackages = new Dictionary<string, HashSet<ResolverPackage>>();
```

Comment:
Add a comment that this should be sorted topologically first, it's not clear within that method that this has been optimized before passing it in